packages/orn-locator/src/resolvers/books.ts

```diff
@@ -208,3 +211,69 @@ export const element = async({bookId, pageId, elementId}: {bookId: string; pageI
     }
   };
 };
+
+const memoryTreeSearch = (getRank: (node: any) => number) => (node: any): Array<{node: any; rank: number}> => {
+  const rank = getRank(node);
+  const contents = ('contents' in node ? node.contents.map(memoryTreeSearch(getRank)) : [])
+    .reduce((result: any[], item: any[]) => [...result, ...item], []);
+
+  return rank > 0 ? [{node, rank}, ...contents] : contents;
+};
+export const bookSearch = async(query: string, filters: {[key: string]: string | string[]} = {}): Promise<Awaited<ReturnType<typeof book>>[]> => {
+  const scopes = 'scope' in filters
+    ? (await locateAll(typeof filters.scope === 'string' ? [filters.scope] : filters.scope))
+      .filter(isResourceOrContentOfTypeFilter(['library']))
+    : [await library('all')];
+
+  return doSearch(query, ['book'])(scopes);
+};
+export const pageSearch = async(query: string, filters: {[key: string]: string | string[]} = {}): Promise<Awaited<ReturnType<typeof page>>[]> => {
+  const scopes = 'scope' in filters
+    ? (await locateAll(typeof filters.scope === 'string' ? [filters.scope] : filters.scope))
+      .filter(isResourceOrContentOfTypeFilter(['book'])).map(book => book.id)
+    : (await library('all')).contents.map(book => book.id);
+
+  return await Promise.all(scopes.map(bookDetail))
+    .then(doSearch(query, ['book:page']))
+  ;
+};
+
+const doSearch = (query: string, filterTypes: string[]) => (inputs: any[]): Promise<any[]> => {
+  const results = inputs.map(memoryTreeSearch(parseSearchQuery(query, filterTypes)))
+    .reduce((result, item) => [...result, ...item], []);
+
+  results.sort((a, b) => b.rank - a.rank);
+  return locateAll(
+    results
+      .slice(0, 5)
+      .map(result => result.node.orn)
+  );
+};
+const parseSearchQuery = (query: string, filterTypes: string[]) => {
+  const quotedTerms = [...query.matchAll(/"(.+)"/g)].map(match => match[1]);
```

Comment:
This should probably be `[^"]+` instead of `.+` otherwise the match is garbled if there are multiple quotes.